### ISSUE TITLE
Integration with spring-boot based projects

### ISSUE THREAD

**Issue by LukaszByczynski** [stagemonitor/stagemonitor#69]
I wonder how enable stagemonitor with spring-boot?

**Comment by felixbarny** [comment]
The current problem is, that the embedded container does not scan for `@WebFilter` and `@WebServlet` annotations. See also https://github.com/spring-projects/spring-boot/issues/2290. I may have to rethink the use of these annotations.

**Comment by yacota** [comment]
I've faced the same problem, the following configuration class defines all the necessary beans.

{code}
@Configuration
public class StageMonitorWebViewConfiguration {
    
    public static final int ORDER = Ordered.HIGHEST_PRECEDENCE+10;
    
    @Bean
    public ServletRegistrationBean rumServlet() {
        ServletRegistrationBean registrationBean = new ServletRegistrationBean();
        RumServlet servlet = new org.stagemonitor.web.monitor.rum.RumServlet();
        registrationBean.setName("stageRumServlet");
        registrationBean.setOrder(ORDER);
        registrationBean.setServlet(servlet);
        registrationBean.addUrlMappings("/stagemonitor/rum");
        return registrationBean;
    }
    
    @Bean
    public ServletRegistrationBean requestTracerServlet() {
        ServletRegistrationBean registrationBean = new ServletRegistrationBean();
        RequestTraceServlet servlet = new org.stagemonitor.web.monitor.widget.RequestTraceServlet();
        registrationBean.setName("stageRequestTracerServlet");
        registrationBean.setOrder(ORDER);
        registrationBean.setServlet(servlet);
        registrationBean.addUrlMappings("/stagemonitor/request-traces");
        return registrationBean;
    }
    
    @Bean
    public ServletRegistrationBean metricServlet() {
        ServletRegistrationBean registrationBean = new ServletRegistrationBean();
        StagemonitorMetricsServlet servlet = new org.stagemonitor.web.metrics.StagemonitorMetricsServlet();
        registrationBean.setName("stageMetricServlet");
        registrationBean.setOrder(ORDER);
        registrationBean.setServlet(servlet);
        registrationBean.addUrlMappings("/stagemonitor/metrics");
        return registrationBean;
    }
    
    @Bean
    public FilterRegistrationBean   requestMonitorFilter() {
        FilterRegistrationBean registrationBean = new FilterRegistrationBean();
        HttpRequestMonitorFilter filter = new HttpRequestMonitorFilter();
        registrationBean.setName("requestMonitorFilter");
        registrationBean.setOrder(ORDER);
        registrationBean.setFilter(filter);
        registrationBean.setUrlPatterns(Arrays.asList("/*"));
        registrationBean.setDispatcherTypes(EnumSet.of(javax.servlet.DispatcherType.REQUEST, javax.servlet.DispatcherType.FORWARD));
        return registrationBean;
    }
    
    @Bean 
    public ServletRegistrationBean configurationServlet() {
        ServletRegistrationBean registrationBean = new ServletRegistrationBean();
        ConfigurationServlet servlet = new ConfigurationServlet();
        registrationBean.setName("stageConfigurationServlet");
        registrationBean.setOrder(ORDER);
        registrationBean.setServlet(servlet);
        registrationBean.addUrlMappings("/stagemonitor/configuration");
        return registrationBean;
    }
}
{code}

Hope it helps

**Comment by felixbarny** [comment]
I think you also need to add the `@WebListener`-cnnotated classes. These are:

```
RuntimeAgentAttacherListener
MDCListener
MonitoredHttpRequest
SpringMonitoredHttpRequest
SessionCounter
```

It really is a pity that these annotations are not picked up automatically. It's not acceptable that stagemonitor users have to add this spring configuration, but at the moment I don't have an idea how to do better.

**Comment by LukaszByczynski** [comment]
You can add separate spring-boot-starter project for stage monitor with all the  configuration classes required to work. Like in [Vaadin addon for spring-boot](https://github.com/vaadin/spring/blob/master/vaadin-spring-boot/src/main/java/com/vaadin/spring/boot/VaadinAutoConfiguration.java) there is ```@ConditionalOnClass``` to provide detection if all required classes exists in classpath.

**Comment by yacota** [comment]
+1 to an artifact stagemonitor-spring-boot

**Comment by felixbarny** [comment]
I'll probably do that, but I would still prefer a solution that does not require any spring specific stuff.

**Comment by LukaszByczynski** [comment]
Separate artifact is well designed solution to keep all that boilerplate code far away from user code.I know, this is not perfect, but all benefits of the spring-boot are worthy of this little piece of sacrifice.

**Comment by felixbarny** [comment]
Here is a example of a workaround: https://github.com/thomasdarimont/spring-boot-jee-web-example

**Comment by felixbarny** [comment]
Ok, here is a better workaround that is working with the just released version 0.15.0:

```
@SpringBootApplication
public class Application implements EmbeddedServletContainerCustomizer {

	public static void main(String[] args) throws Exception {
		SpringApplication.run(SampleWebUiApplication.class, args);
	}

	@Override
	public void customize(ConfigurableEmbeddedServletContainer container) {
		container.addInitializers(new ServletContextInitializer() {
			@Override
			public void onStartup(ServletContext servletContext) throws ServletException {
				new WebPlugin().onStartup(null, servletContext);
				// uncomment if you are using the alerting plugin
				// new AlertingPlugin().onStartup(null, servletContext);
			}
		});
	}
}
```

Sorry for the slow turnaround.

I'm also working with the guys from spring boot on a solution that does not require any code changes to your application (see https://github.com/spring-projects/spring-boot/issues/321). Note that this workaround is only required, if you are not creating a war that is deployed on a servlet container.

**Comment by LukaszByczynski** [comment]
Thank you for support and workaround.
